cpp/include/cuspatial/experimental/iterator_factory.cuh

```diff
@@ -54,6 +75,74 @@ struct vec_2d_to_tuple {
   }
 };
 
+/**
+ * @internal
+ * @brief Generic to convert any iterator pointing to interleaved xy range into
+ * iterator of vec_2d.
+ *
+ * This generic version does not make use of vectorized load.
+ *
+ * @pre `Iter` has operator[] defined.
+ * @pre std::iterator_traits<Iter>::value_type is convertible to `T`.
+ */
+template <typename Iter, typename Enable = void>
+struct interleaved_to_vec_2d {
+  using T = typename std::iterator_traits<Iter>::value_type;
+  Iter it;
+  __device__ vec_2d<T> operator()(std::size_t i) { return vec_2d<T>{it[2 * i], it[2 * i + 1]}; }
+};
+
+/**
+ * @brief Specialization for raw pointers to interleaved xy range.
+ *
+ * @pre `Iter` is raw pointer
+ */
+template <typename Iter>
+struct interleaved_to_vec_2d<Iter, typename std::enable_if_t<std::is_pointer_v<Iter>>> {
+  using pointer_t = typename std::pointer_traits<Iter>::pointer;
+  using T         = typename std::remove_const_t<typename std::pointer_traits<Iter>::element_type>;
+  using T2        = typename element_to_element2<T>::type;
+  pointer_t ptr;
+
+  __device__ vec_2d<T> operator()(std::size_t i)
+  {
+    T2 const* f2it = reinterpret_cast<T2 const*>(ptr);
```

Comment:
Hmmm, I think this is technically UB.  I think it requires `std::memcpy` (or `std::bitcast` after C++20).  Does it work if you use `std::memcpy`?  If so, assuming the alignment of `vec_2d<T>` is the same as `T2` (float2, double2), then why not copy directly to a vec_2d rather than a T2?

```suggestion
    vec_2d<T> v2;
    std::memcpy(&v2, ptr, sizeof(v2));
    return v2;
```